### ISSUE TITLE
[heft] heft does not seem to be compiling incrementally

### ISSUE THREAD

**Issue by scamden** [microsoft/rushstack#2588]
i've created a simple repro here: https://github.com/creditiq/simple-repros/tree/heft-incremental-test

Repro: 
1. clone the above
2. `cd simple-repros/dep`
3. `npm install`
4. `heft build --verbose`
5. `heft build --verbose`

Expected:
second build has 0 check and emit time (this is the case if running a direct tsc --diagnostics)

Actual:
All logged timings seem the same

(i'll comment below with my outputs in case you don't want to run the steps above yourself)

**Comment by scamden** [comment]
[heft build verbose.txt](https://github.com/microsoft/rushstack/files/6267240/heft.build.verbose.txt)
[tsc diagnostics.txt](https://github.com/microsoft/rushstack/files/6267241/tsc.diagnostics.txt)

**Comment by scamden** [comment]
i've also noticed that heft has some significant overhead compared to basic tsc. 

you can see in the attached files that the overall non-incremental compile took .4s longer with heft. when i tried this on a larger code base the effect was even greater with heft taking around 3 seconds longer than basic typescript compile. so it seems like the overhead does scale somewhat with the code size and isn't just configuration and setup etc like i wondered. 

(maybe this is just due to extra logging io or something since i'm in verbose mode, but I thought I should mention it)

**Comment by dmichon-msft** [comment]
Heft will be slightly slower than `tsc`, because Heft uses the TypeScript Compiler API, which is internally slightly less optimized than tsc, especially with respect to the AST representation.
The incremental check, however, is a bug, though a difficult one to track down.

**Comment by octogonz** [comment]
> which is internally slightly less optimized than tsc, especially with respect to the AST representation.

Do you have any details about that?  `typescript/lib/tsc.js` appears to be bundled differently from `typescript/lib/typescript.js`, but is there some way for Heft to fix this, for example by Webpacking some modules with optimizations enabled?

**Comment by scamden** [comment]
@dmichon-msft thanks for the info on the overhead. it's definitely not ideal but tolerable. although i would be interested about @octogonz 's ideas to improve that. 3 seconds is 3 seconds. 

> The incremental check, however, is a bug, though a difficult one to track down.

sorry :P if it helps one thing i've noticed is that the buildinfo files tsc generates have `"signature": <hash>` entries for every file but the one created by heft doesn't. not sure if that's much of lead but maybe..

**Comment by dmichon-msft** [comment]
Internally tsc uses a different construction of AST nodes, since it doesn't have to support the public API that tools like tslint consume. The construction is a little more efficient in the v8 runtime due to its memory representation.

**Comment by scamden** [comment]
@dmichon-msft so the trade off is basically, a sunk cost of that overhead to gain the win of not having to compile twice for two different emit targets? are there other gains heft gets for using that api? if not could it swap in tsc when not doing two emit targets?

**Comment by dmichon-msft** [comment]
Technically we could reroute the AST node allocator to use the more efficient one, but I suspect that portion of the cost is actually negligible here. More likely is the bug with the incremental cost, as well as boot up overhead because Heft loads more than just TypeScript. The savings of using the the compiler service AST allocator comes from being able to share the Program instance with the linter, but eslint currently doesn't take advantage of that.

**Comment by scamden** [comment]
@dmichon-msft i was wondering about boot up overhead but i think it must be more than that because the overhead scales with the size of the project. maybe something about heft's file io or something? the incremental is def significant, but i think it's a separate issue. even comparing two clean compiles heft was 3 seconds slower on a large package, .5 on a small one. seems like it must be some overhead that scales with the number of files or the amount of code..

**Comment by scamden** [comment]
i'll open a separate issue for the overhead issue actually. the incremental thing seems more fixable and is a much larger time savings for us anyway.

**Comment by dmichon-msft** [comment]
Well, you will definitely get scaling cost from the dual emit (if you are using it), since the transform and emit is actually one of the more expensive parts of TypeScript (at least in our repos). The savings comes from type checking and parsing only once.

**Comment by scamden** [comment]
oh dang that's a really good point. let me see if it even scales on that larger repo if i disable the second output.

**Comment by scamden** [comment]
ah geez i feel silly. they are very close together without the additional emit. i really should have seen that apologies for the wasted time. i'll close the other issue. looks like the only overhead is the very small startup cost.

**Comment by dmichon-msft** [comment]
Found the incremental issue, at least in my local builds:
Ensure that `noEmitOnError` is set to `false` in your tsconfig, otherwise TypeScript does a full type-check pass every run for the declaration file diagnostics.

**Comment by dmichon-msft** [comment]
Still something else going on in Heft; I'm not able to get it to build incrementally at all (not just type check).

**Comment by dmichon-msft** [comment]
Found root cause, fix here: #2612

**Comment by dmichon-msft** [comment]
Fixed in #2612

**Comment by scamden** [comment]
```
[typescript] Check: 0ms
[typescript] Transform: 0ms (0 files)
[typescript] Print: 0ms (0 files) (Includes Transform)
[typescript] Emit: 0ms (Includes Print)
```
💥 thank you! 💥

**Comment by octogonz** [comment]
Awesome!!
